helix-core/src/main/java/org/apache/helix/controller/rebalancer/topology/Topology.java

```diff
@@ -230,20 +199,73 @@ private boolean isInstanceEnabled(ClusterConfig clusterConfig, String instanceNa
   }
 
+  /**
+   * Populate faultZone, endNodetype and and a LinkedHashMap containing pathKeys default values for
+   * clusterConfig.Topology. The LinkedHashMap will be empty if clusterConfig.Topology is unset.
+   *
+   * @return an Instance of Topology.ClusterTopologyConfig.
+   */
+  private static ClusterTopologyConfig getClusterTopologySetting(ClusterConfig clusterConfig) {
+
+    ClusterTopologyConfig clusterTopologyConfig = new ClusterTopologyConfig();
+    if (clusterConfig.isTopologyAwareEnabled()) {
+      String topologyDef = clusterConfig.getTopology();
+      if (topologyDef != null) {
+        String[] topologyKeys = topologyDef.trim().split("/");
+        int lastValidTypeIdx = 0;
+        for (int i = 0; i < topologyKeys.length; i++) {
+          if (topologyKeys[i].length() != 0) {
+            clusterTopologyConfig.topologyKeyDefaultValue
+                .put(topologyKeys[i], DEFAULT_DOMAIN_PREFIX + topologyKeys[i]);
+            lastValidTypeIdx = i;
+          }
+        }
+        if (clusterTopologyConfig.topologyKeyDefaultValue.size() == 0) {
+          throw new IllegalArgumentException("Invalid cluster topology definition " + topologyDef);
+        }
+        clusterTopologyConfig.endNodeType = topologyKeys[lastValidTypeIdx];
+        String faultZoneType = clusterConfig.getFaultZoneType();
+        if (faultZoneType == null) {
+          clusterTopologyConfig.faultZoneType = clusterTopologyConfig.endNodeType;
+        } else if (!clusterTopologyConfig.topologyKeyDefaultValue.containsKey(faultZoneType)) {
+          throw new HelixException(String
+              .format("Invalid fault zone type %s, not present in topology definition %s.",
+                  faultZoneType, clusterConfig.getTopology()));
+        } else {
+          clusterTopologyConfig.faultZoneType = faultZoneType;
+        }
+      } else {
+        // Use default cluster topology definition, i,e. /root/zone/instance
+        clusterTopologyConfig.endNodeType = Types.INSTANCE.name();
+        clusterTopologyConfig.faultZoneType = Types.ZONE.name();
+      }
+    } else {
+      clusterTopologyConfig.endNodeType = Types.INSTANCE.name();
+      clusterTopologyConfig.faultZoneType = Types.INSTANCE.name();
+    }
+    return clusterTopologyConfig;
+  }
+
   /**
    * This function returns a LinkedHashMap<String, String> object representing
    * the topology path for an instance.
    * LinkedHashMap is used here since the order of the path needs to be preserved
    * when creating the topology tree.
    *
+   * @param clusterTopologyKeyDefaultValue  a LinkedHashMap where keys are cluster topology path and
+   *                                       values are their corresponding default value. The entries
+   *                                        are ordered by ClusterConfig.topology setting.
+   * @param faultZoneForEarlyQuit   this flag is set to true only if caller wants the path
```

Comment:
This comment should be for `computeInstanceTopologyMap()`? That public function needs an entire javadoc also.

Reply:
TFTR. Updated.